Use std_{code,expr,type} constructors

tautschnig opened the pull request:
Applies resource-acquisition-is-initialisation. This might slightly improve
performance, but otherwise should not change behaviour. While it will ensure
consistent data structures, no bugs were spotted/fixed here.

kroening commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -633,10 +633,10 @@ bool acceleration_utilst::do_arrays(
-  exprt forall(ID_forall);
+  forall_exprt forall;
   forall.type()=bool_typet();
-  forall.copy_to_operands(k);
```
forall is always boolean, so I'd be tempted to tweak the class definition of forall_exprt to derive from binary_predicate_exprt instead of binary_exprt.

kroening replied:
Oh, and also to add a constructor that takes two arguments.

kroening commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -785,10 +784,10 @@ bool acceleration_utilst::do_arrays(
-    exprt array_unchanged(ID_forall);
+    forall_exprt array_unchanged;
     array_unchanged.type()=bool_typet();
-    array_unchanged.copy_to_operands(idx);
```
same here

kroening commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -1080,10 +1079,10 @@ bool acceleration_utilst::assign_array(
-  exprt forall(ID_forall);
+  forall_exprt forall;
   forall.type()=bool_typet();
-  forall.copy_to_operands(k);
```
and again

kroening commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -1070,7 +1069,7 @@ bool acceleration_utilst::assign_array(
   symbolt k_sym=fresh_symbol("polynomial::k", unsigned_poly_type());
-  exprt k=k_sym.symbol_expr();
+  symbol_exprt k = k_sym.symbol_expr();
 
```
I would make that one const, which makes it clearer that 'k' later on doesn't change

kroening commented on src/goto-instrument/accelerate/polynomial_accelerator.cpp:
```diff
@@ -223,10 +223,10 @@ bool polynomial_acceleratort::accelerate(
-    exprt forall(ID_forall);
+    forall_exprt forall;
     forall.type()=bool_typet();
-    forall.copy_to_operands(k);
```
another one

kroening commented on src/solvers/qbf/qbf_bdd_core.cpp:
```diff
@@ -285,12 +285,8 @@ const exprt qbf_bdd_certificatet::f_get(literalt l)
-    extract_expr.copy_to_operands(from_integer(index, uint_type));
+    extractbit_exprt extract_expr(
+      sym, from_integer(index, unsignedbv_typet(32)));
 
```
Simply use integer_typet()?

kroening commented on src/solvers/qbf/qbf_squolem_core.cpp:
```diff
@@ -198,11 +198,8 @@ const exprt qbf_squolem_coret::f_get(literalt l)
-    extract_expr.copy_to_operands(from_integer(index, uint_type));
+    extractbit_exprt extract_expr(
+      sym, from_integer(index, unsignedbv_typet(32)));
 
```
Same here.

kroening reviewed (approved):
Only minor tweaks, the forall tweaks could be done separately.

thk123 commented on src/analyses/invariant_propagation.cpp:
```diff
@@ -100,8 +100,8 @@ void invariant_propagationt::get_objects_rec(
 
-    exprt member_expr(ID_member);
-    member_expr.copy_to_operands(src);
+    member_exprt member_expr;
```
No need to do it as part of this PR, but seems like would make more sense to introduce this _inside_ the for loop then you can construct from the `component` directly.

thk123 replied:
But you should probably use `set_component_name`  on line 108 here

thk123 commented on src/analyses/invariant_set.cpp:
```diff
@@ -491,10 +491,10 @@ void invariant_sett::strengthen_rec(const exprt &expr)
-      lhs_member_expr.copy_to_operands(expr.op0());
-      rhs_member_expr.copy_to_operands(expr.op1());
+      member_exprt lhs_member_expr;
+      member_exprt rhs_member_expr;
```
Use `set_component_name` on the lines below

thk123 commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1443,8 +1439,8 @@ void c_typecheck_baset::typecheck_expr_ptrmember(exprt &expr)
 
-  exprt deref(ID_dereference);
-  deref.move_to_operands(expr.op0());
+  dereference_exprt deref;
```
`dereference_exprt deref{expr.op0()}`?

thk123 replied:
Or perhaps since this doesn't actually set the type, perhaps `nil_typet` or something to precisely mimic existing behaviour?

thk123 commented on src/cpp/cpp_typecheck_constructor.cpp:
```diff
@@ -669,6 +669,8 @@ void cpp_typecheckt::full_member_initialization(
 
     if(!vbases.empty())
     {
+      // TODO: this code doesn't seem to make much sense as the ifthenelse
```
Nit annotated the TODO with your name `TODO(mtautschnig): `

thk123 commented on src/cpp/cpp_typecheck_constructor.cpp:
```diff
@@ -792,6 +794,8 @@ void cpp_typecheckt::full_member_initialization(
 
       if(parent_it->get_bool(ID_virtual))
       {
+        // TODO: this code doesn't seem to make much sense as the ifthenelse
```
Nit annotated the TODO with your name `// TODO(mtautschnig):`

thk123 commented on src/cpp/cpp_typecheck_destructor.cpp:
```diff
@@ -159,8 +159,8 @@ codet cpp_typecheckt::dtor(const symbolt &symbol)
 
-    exprt object(ID_dereference);
-    object.operands().push_back(exprt("cpp-this"));
+    dereference_exprt object;
```
Should be able to use the  constructor for setting `op` here?

thk123 reviewed (commented):
Got to `src/goto-instrument/accelerate/polynomial_accelerator.cpp` if someone wants to pick up from there, but looks like a nice clean up!

tautschnig committed:
Use std_{code,expr,type} constructors

Applies resource-acquisition-is-initialisation. This might slightly improve
performance, but otherwise should not change behaviour. While it will ensure
consistent data structures, no bugs were spotted/fixed here.

Added constructors to {forall,exists,quantifier}_exprt two operands to enable
RAII.

tautschnig commented:
@thk123 @kroening Thanks a lot for all the comments, all addressed.

tautschnig commented:
I note that this should go in after #1897 (and it will require a conflict-resolving rebase at that time).

martin-cs reviewed (approved):
I agree with the general idea; irept's becoming more functional is not a bad thing.  I'm concerned that this kind of wide change, even if localised, will break external patch sets.  Hearing from a few people that it wouldn't would help me rest more easily in that regard.

smowton commented:
Don't really have time to read this in detail today, sorry, so I'm going to duck in favour of another code owner

romainbrenguier commented on src/analyses/goto_check.cpp:
```diff
@@ -228,8 +228,7 @@ void goto_checkt::div_by_zero_check(
 
-  exprt inequality(ID_notequal, bool_typet());
-  inequality.copy_to_operands(expr.op1(), zero);
+  notequal_exprt inequality(expr.op1(), zero);
```
const

romainbrenguier commented on src/analyses/goto_check.cpp:
```diff
@@ -329,8 +328,7 @@ void goto_checkt::mod_by_zero_check(
 
-  exprt inequality(ID_notequal, bool_typet());
-  inequality.copy_to_operands(expr.op1(), zero);
+  notequal_exprt inequality(expr.op1(), zero);
```
const

romainbrenguier commented on src/analyses/goto_check.cpp:
```diff
@@ -374,13 +372,13 @@ void goto_checkt::conversion_check(
-        binary_relation_exprt no_overflow_upper(ID_le);
-        no_overflow_upper.lhs()=expr.op0();
-        no_overflow_upper.rhs()=from_integer(power(2, new_width-1)-1, old_type);
+        binary_relation_exprt no_overflow_upper(
```
const here and on all the other modified lines of this file

romainbrenguier commented on src/ansi-c/ansi_c_entry_point.cpp:
```diff
@@ -270,8 +270,7 @@ bool generate_ansi_c_start_function(
 
-        exprt ge(ID_ge, typet(ID_bool));
-        ge.copy_to_operands(argc_symbol.symbol_expr(), one);
+        binary_relation_exprt ge(argc_symbol.symbol_expr(), ID_ge, one);
```
const ? (applies to the other changes in this file)

romainbrenguier commented on src/cbmc/bv_cbmc.cpp:
```diff
@@ -79,50 +69,27 @@ bvt bv_cbmct::convert_waitfor(const exprt &expr)
+
+      and_exprt and_expr(cycle_le_bound, cycle_eq_new_cycle);
+
+      implies_exprt top_impl(and_expr, tmp_predicate);
```
You can remove the blank lines and add const everywhere

romainbrenguier commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -1070,7 +1062,7 @@ bool acceleration_utilst::assign_array(
 
   symbolt k_sym=fresh_symbol("polynomial::k", unsigned_poly_type());
-  exprt k=k_sym.symbol_expr();
+  const symbol_exprt k = k_sym.symbol_expr();
```
can it be a reference?

tautschnig replied:
No, `symbolt::symbol_expr` returns by value.

romainbrenguier commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -1070,7 +1062,7 @@ bool acceleration_utilst::assign_array(
+  const symbol_exprt k = k_sym.symbol_expr();
 
   exprt k_bound=
     and_exprt(
```
`const and_exprt k_bound(`

romainbrenguier commented on src/solvers/floatbv/float_bv.cpp:
```diff
@@ -635,7 +635,8 @@ exprt float_bvt::mul(
   {
-    exprt NaN_cond(ID_or, bool_typet());
+    or_exprt NaN_cond;
+    NaN_cond.reserve_operands(4);
```
Could use `disjunction` instead

thk123 reviewed (approved):
Remembered only need to review `analyses` and `java_bytecode` - these look fine (and previous suggestions have been applied) so happy for this to go in

tautschnig commented:
I've just found another 250 locations that may be amenable to this sort of improvement, I'll post updates later on today.

martin-cs commented:
On Fri, 2018-03-02 at 10:48 +0000, Michael Tautschnig wrote:
> I've just found another 250 locations that may be amenable to this sort of improvement, I'll post updates later on today.

<expression of concern about the size of the resultant change and the
impact on other patches>

tautschnig commented:
> expression of concern about the size of the resultant change and the impact on other patches

Concern noted and shared; but also those changes seem to be the right-thing-to-do(TM). Does it help in any way that the code patched here typically hasn't been touched in a long time as otherwise it wouldn't have been written in this style in the first place?

kroening commented:
I am positive about these sort of changes; they clearly improve readability, maintainability and robustness. Will merge once CIed.
